README.md
# RRule

RRule is a wrapper around the Rust based library [rrule](https://github.com/fmeringdal/rust-rrule).

To consume it in your projects, you'll need to have a Rust compiler [installed](https://www.rust-lang.org/en-US/install.html).

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
by adding `rrule` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:rrule, "~> 0.1.0"}
  ]
end
```

CHANGELOG.md
## 0.5.2

- Bumps Rustler to support OTP 26 and updates chrono to `0.8.1` [#1](https://github.com/mfeckie/rrule/pull/1)

## 0.6.0

- Switches to rustler precompiled so you don't need to have a rust compiler installed to use the library if you're on an architecture supported by the precompiler :tada:

## 0.6.1

- Bump Rustler

test/rrule_test.exs
defmodule RRuleTest do
  use ExUnit.Case, async: true
  doctest RRule

  test "Lists occurrences between for RRULE" do
    {:ok, {occurrences, has_more}} =
      RRule.all_between(
        "DTSTART:20120101T093000Z\nRRULE:FREQ=DAILY;COUNT=5",
        ~U[2012-01-01 09:00:00Z],
        ~U[2012-02-01 09:00:00Z]
      )

    assert occurrences == [
             ~U[2012-01-01 09:30:00Z],
             ~U[2012-01-02 09:30:00Z],
             ~U[2012-01-03 09:30:00Z],
             ~U[2012-01-04 09:30:00Z],
             ~U[2012-01-05 09:30:00Z]
           ]

    refute has_more
  end

  test "Lists occurrences between for RRULESET" do
    {:ok, {occurrences, has_more}} =
      RRule.all_between(
        ~s(DTSTART:20120101T093000Z\nRRULE:FREQ=MONTHLY;COUNT=100\nRDATE:20120201T023000Z,20120702T023000Z\nEXRULE:FREQ=MONTHLY;COUNT=2\nEXDATE:20120601T023000Z),
        ~U[2012-01-01 09:00:00Z],
        ~U[2012-12-01 09:00:00Z],
        2
      )

    assert occurrences == [
             ~U[2012-01-01 09:30:00Z],
             ~U[2012-02-01 02:30:00Z]
           ]

    assert has_more
  end

  test "Reports error for all_between when rule can't be parsed" do
    {:error, msg} =
      RRule.all_between(
        "DTSTART:2012010=DAILY;COUNT=5",
        ~U[2012-01-01 09:00:00Z],
        ~U[2012-02-01 09:00:00Z]
      )

    assert msg ==
             "RRule parsing error: `2012010=DAILY;COUNT=5` is not a valid datetime format for `DTSTART`."
  end

  test "Validates RRULE" do
    assert :ok ==
             RRule.validate("DTSTART:20120101T093000Z\nRRULE:FREQ=DAILY;COUNT=5")
  end

  test "Error for invalid RRULE" do
    assert {:error,
            "RRule parsing error: `DTSTA` is not a valid property name, expected one of: `RRULE,EXRULE,DTSTART,RDATE,EXDATE`"} ==
             RRule.validate("DTSTA:20120101T093000Z\nRRULE:FREQ=DAILY;COUNT=5")
  end

  test "Retrieve DTSTART for RRULE" do
    {:ok, start_date} = RRule.get_start_date("DTSTART:20120101T093000Z\nRRULE:FREQ=DAILY;COUNT=5")

    assert start_date == ~U[2012-01-01 09:30:00Z]
  end

  test "Lists all occurrences for RRULE up to limit" do
    {:ok, occurrences} =
      RRule.all(
        "DTSTART:20120101T093000Z\nRRULE:FREQ=DAILY;COUNT=100",
        10
      )

    assert occurrences == [
             ~U[2012-01-01 09:30:00Z],
             ~U[2012-01-02 09:30:00Z],
             ~U[2012-01-03 09:30:00Z],
             ~U[2012-01-04 09:30:00Z],
             ~U[2012-01-05 09:30:00Z],
             ~U[2012-01-06 09:30:00Z],
             ~U[2012-01-07 09:30:00Z],
             ~U[2012-01-08 09:30:00Z],
             ~U[2012-01-09 09:30:00Z],
             ~U[2012-01-10 09:30:00Z]
           ]
  end

  test "Returns error tuple when limit is exceeded" do
    {:error, msg} =
      RRule.all(
        "DTSTART:20120101T093000Z\nRRULE:FREQ=DAILY;COUNT=100",
        65_536
      )

    assert msg == "Limit must be below 65,535"
  end
end


mix.exs
defmodule RRule.MixProject do
  use Mix.Project

  @version "0.6.1"

  def project do
    [
      app: :rrule,
      version: @version,
      elixir: "~> 1.13 or ~> 1.14",
      description: "Elixir wrapper for Rust based RRule parsing",
      start_permanent: Mix.env() == :prod,
      package: package(),
      docs: [
        extras: ["README.md"],
        main: "readme"
      ],
      deps: deps()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  def package do
    [
      licenses: ["MIT"],
      maintainers: ["Martin Feckie"],
      links: %{
        "Github" => "https://github.com/mfeckie/rrule"
      },
      files: ["lib", "native", "README.md", "mix.exs", "checksum-*.exs"],
      exclude_patterns: [
        "native/rrule/target"
      ]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:rustler, "~> 0.29.0"},
      {:rustler_precompiled, "0.6.1"},
      {:ex_doc, "~> 0.29.2"}
    ]
  end
end


native/rrule/src/lib.rs
extern crate rustler;
use chrono::{DateTime, Datelike, TimeZone, Timelike};
use rrule::{RRuleSet, Tz};
use rustler::{Atom, Encoder, Env, NifStruct, Term};

#[rustler::nif]
fn parse<'a>(env: Env<'a>, string: &str) -> Result<Term<'a>, String> {
    // let rrule: RRuleSet = string.parse().unwrap();

    // (rrule.to_string()).encode(env)
    let parsed: String = match string.parse::<RRuleSet>() {
        Ok(rrule) => rrule.to_string(),
        Err(err) => return Err(format!("{}", err)),
    };

    Ok((parsed).encode(env))
}

mod atoms {
    rustler::atoms! {
        calendar_iso = "Elixir.Calendar.ISO"
    }
}

// __struct__: DateTime,
// calendar: Calendar.ISO,
// day: 23,
// hour: 1,
// microsecond: {490236, 6},
// minute: 23,
// month: 7,
// second: 15,
// std_offset: 0,
// time_zone: "Etc/UTC",
// utc_offset: 0,
// year: 2022,
// zone_abbr: "UTC"
#[derive(Debug, NifStruct)]
#[module = "DateTime"]
struct ElixirDateTime {
    calendar: Atom,
    day: u32,
    month: u32,
    year: i32,
    hour: u32,
    minute: u32,
    second: u32,
    microsecond: (i32, i32),
    std_offset: i32,
    utc_offset: i32,
    time_zone: String,
    zone_abbr: String,
}

impl ElixirDateTime {
    pub fn new(input: &DateTime<Tz>) -> ElixirDateTime {
        ElixirDateTime {
            calendar: atoms::calendar_iso(),
            day: input.naive_utc().day(),
            month: input.naive_utc().month(),
            year: input.naive_utc().year(),
            hour: input.naive_utc().hour(),
            minute: input.naive_utc().minute(),
            second: input.naive_utc().second(),
            microsecond: (0, 0),
            std_offset: 0,
            utc_offset: 0,
            time_zone: "Etc/UTC".to_string(),
            zone_abbr: "UTC".to_string(),
        }
    }

    fn to_chrono(&self) -> DateTime<Tz> {
        Tz::UTC
            .with_ymd_and_hms(self.year, self.month, self.day, self.hour, self.minute, self.second)
            .unwrap()
    }
}

#[rustler::nif]
fn all<'a>(env: Env<'a>, string: &str, limit: u64) -> Result<Term<'a>, String> {
    if limit > 65535 {
        return Err(format!("Limit must be below 65,535"))
    }

    let rrule: RRuleSet = match string.parse() {
        Ok(parsed) => parsed,
        Err(err) => return Err(format!("{}", err)),
    };

    let (results, _has_more) = rrule
    .all(limit as u16);

    let mapped: Vec<ElixirDateTime> = results.iter().map(ElixirDateTime::new).collect();

    Ok((mapped).encode(env))
}

#[rustler::nif]
fn all_between<'a>(
    env: Env<'a>,
    string: &str,
    start_date: ElixirDateTime,
    end_date: ElixirDateTime,
    limit: u64,
) -> Result<Term<'a>, String> {
    if limit > 65535 {
        return Err(format!("Limit must be below 65,535"))
    }

    let rrule: RRuleSet = match string.parse() {
        Ok(parsed) => parsed,
        Err(err) => return Err(format!("{}", err)),
    };

    let (results, has_more) = rrule
        .after(start_date.to_chrono())
        .before(end_date.to_chrono())
        .all(limit as u16);

    let mapped: Vec<ElixirDateTime> = results.iter().map(ElixirDateTime::new).collect();
    Ok(((mapped, has_more)).encode(env))
}

#[rustler::nif]
fn validate<'a>(env: Env<'a>, rrule: &str) -> Term<'a> {
    match rrule.parse::<RRuleSet>() {
        Ok(_parsed) => (rustler::types::atom::ok()).encode(env),
        Err(err) => (rustler::types::atom::error(), format!("{}", err)).encode(env),
    }
}

#[rustler::nif]
fn get_start_date<'a>(env: Env<'a>, rrule: &str) -> Result<Term<'a>, String> {
    let parsed_rule = match rrule.parse::<RRuleSet>() {
        Ok(parsed) => parsed,
        Err(err) => return Err(format!("{}", err)),
    };

    let start_date_time = ElixirDateTime::new(parsed_rule.get_dt_start());

    Ok((start_date_time).encode(env))
}

rustler::init!(
    "Elixir.RRule",
    [all, all_between, get_start_date, parse, validate]
);


lib/rrule.ex
defmodule RRule do
  mix_config = Mix.Project.config()
  version = mix_config[:version]

  use RustlerPrecompiled,
    otp_app: :rrule_ex,
    crate: :rrule_ex,
    base_url: "https://github.com/mfeckie/rrule/releases/download/#{version}",
    force_build: System.get_env("RUSTLER_PRECOMPILATION_EXAMPLE_BUILD") in ["1", "true"],
    targets: ~w(
      aarch64-apple-darwin
      aarch64-unknown-linux-gnu
      aarch64-unknown-linux-musl
      riscv64gc-unknown-linux-gnu
      x86_64-apple-darwin
      x86_64-pc-windows-msvc
      x86_64-pc-windows-gnu
      x86_64-unknown-linux-gnu
      x86_64-unknown-linux-musl
    ),
    version: version

  @moduledoc """
  Wrapper functions for interacting with the Rust based rrule library
  """

  @spec parse(String.t()) :: String.t()
  @doc """
  Given a RRule, returns the expanded representation if parsed successfully

  ## Example

      iex> RRule.parse( "DTSTART:20120101T093000Z\\nRRULE:FREQ=DAILY;COUNT=5")
      {:ok, "DTSTART:20120101T093000Z\\nFREQ=daily;COUNT=5;BYHOUR=9;BYMINUTE=30;BYSECOND=0"}
  """
  def parse(_RRule_string), do: :erlang.nif_error(:not_loaded)

  @doc """
  Given a RRule, return occurrences until a limit is reached

  *Note* Maximum occurrences is limited to 65,535 to avoid order to prevent infinite loops

  """
  @spec all(String.t(), non_neg_integer()) :: {:ok, [DateTime.t()]} | {:error, String.t()}
  def all(_rrule_string, _number), do: :erlang.nif_error(:not_loaded)

  @doc """
  Given a RRule, return all occurrences between the given UTC DateTimes.

  In the success case, the return will be `{:ok, {occurrences, has_more }}`.  The `has_more` boolean is used to inform you if there
  are more occurrences that weren't returned due to the limits.  Generally you'll only care about the `occurrences`

  *Note* Maximum occurrences is limited to 65,535 to avoid order to prevent infinite loops

  ## Example
      iex> RRule.all_between("DTSTART:20120101T093000Z\\nRRULE:FREQ=DAILY;COUNT=5", ~U[2012-01-01 09:00:00Z], ~U[2012-02-01 09:00:00Z])
      {:ok,
      {[~U[2012-01-01 09:30:00Z], ~U[2012-01-02 09:30:00Z], ~U[2012-01-03 09:30:00Z],
      ~U[2012-01-04 09:30:00Z], ~U[2012-01-05 09:30:00Z]], false}}
  """
  @spec all_between(String.t(), DateTime.t(), DateTime.t()) ::
          {:ok, {[DateTime.t()], boolean()}} | {:error, String.t()}
  def all_between(string, start_date, end_date) do
    all_between(string, start_date, end_date, 65_535)
  end

  @doc """
  Given a RRule, return n occurrences between the given UTC DateTimes, where n is the limit

  *Note* Limit cannot be greater than 65,535
  """
  @spec all_between(String.t(), DateTime.t(), DateTime.t(), integer()) ::
          {:ok, {[DateTime.t()], boolean()}} | {:error, String.t()}
  def all_between(_RRule_string, _start_date, _end_date, _limit),
    do: :erlang.nif_error(:not_loaded)

  @doc """
  Given a RRule string, return the UTC DateTime when the rule starts

  ## Example
      iex> RRule.get_start_date("DTSTART:20120101T093000Z\\nRRULE:FREQ=DAILY;COUNT=5")
      {:ok, ~U[2012-01-01 09:30:00Z]}
  """
  @spec get_start_date(String.t()) :: {:ok, DateTime.t()} | {:error, String.t()}
  def get_start_date(_rrule_RRule_string), do: :erlang.nif_error(:not_loaded)

  @doc """
  Given a RRule, return a boolean indicating if the rule is valid

      iex> RRule.validate( "DTSTART:20120101T093000Z\\nRRULE:FREQ=DAILY;COUNT=5")
      :ok

      iex> RRule.validate( "DTSTART:20120101T0930Z\\nRRULE:FREQ=DAILY;COUNT=5")
      {:error,
      "RRule parsing error: `20120101T0930Z` is not a valid datetime format for `DTSTART`."}
  """
  @spec validate(String.t()) :: :ok | {:error, String.t()}
  def validate(_RRule_string), do: :erlang.nif_error(:not_loaded)
end
